rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,556 @@
+
+
+### `GET_TELEMETRY`
+
+The request should not have payload.
+
+The response might have payload if the SDK has metrics to report.
+
+The SDK must reset local counters after successful scheduling response for the
+network delivery.
+
+The format of the response follows the [Text-based exposition format of the
+Prometheus](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format),
+and includes metrics and histograms collected by the SDK since the previous
+report.
+
+#### Metrics
+
+    sdk_kv_r_timedout{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a",bucket="travel-sample"} 1
+
+Where
+
+* `sdk_kv_r_timedout` is a name of the metric, one of the:
+
+  | Metric | Description |
+  | ---- | ---- |
+  | `sdk_{service}_r_timedout` | Timeout |
+  | `sdk_{service}_r_canceled` | Canceled |
+  | `sdk_{service}_r_total` | Total number of operations. Note that this number might differ from `_count` of the histogram, if the response from the server has never been received. |
+
+* `{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a",bucket="travel-sample"}` is a list of the tags. The server is free to strip/transform any tags before relaying to Prometheus collector.
+
+  | Label | Description |
+  | ---- | ---- |
+  | `agent` | agent string of the SDK, as it is used everywhere (HELLO message, logs etc.) |
+  | `node` | The hostname (without port) of the node as seen in `nodesExt[].hostname` of the configuration. |
+  | `alt_node` | The hostname (without port)  of the node as seen in `nodesExt[]["alternateAddresses"]["<network>"]["hostname"]`. |
+  | `node_uuid` | The UUID of the node as it is seen in `nodesExt[].nodeUUID` of the configuration. |
+  | `bucket` | Name of the bucket (if present) |
+
+* `1` is a metric value. Each metric value increment represents a single
+  network operation (e.g. `GET_ALL_REPLICAS` should be counted as `1 +
+  number_of_replicas`). The value must be integer (Note that it is more
+  restrictive than Prometheus specification, which allows floats for metric
+  values, but using integers makes it easier for the server to implement
+  aggregation).
+
+Note: while Prometheus allows to specify timestamp in the sample, the SDK do
+not need to include it as the Server will erase it anyway, and is not going to
+use it as the metrics will be aggregated anyway. The EBNF of the format from
+the Prometheus reference.
+
+    metric_name [
+      "{" label_name "=" `"` label_value `"` { "," label_name "=" `"` label_value `"` } [ "," ] "}"
+    ] value [ timestamp ]
+
+#### Histograms
+
+    sdk_kv_retrieval_duration_ms_bucket{le="0.001",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 24054
```

Comment:
The metric name says it's in milliseconds, but the `le` tag is still in seconds. Should we update the `le` values to match the new time unit?

Reply:
Fixed